### DISCUSSION TITLE
Can't launch .app but exec works fine

### DISCUSSION THREAD

**Discussion by web-mojo** [pyinstaller/pyinstaller#7966] [Help] [answered]
Hi all,

I spent 2 days trying to find a solution for my problem. Without success. I've checked the posts of users with similar issues but still couldn't get it to work.

The issue:
I've created a Mac .app file with pyinstaller (--noconsole). When trying to launch the .app the program attempts to launch but immediately closes after a second. I've enabled debugging but even when trying to open the .app from the Terminal, same problem. It opens but closes immediately. No errors in the terminal. I also wrapped the entire code in a try, except statement. Nothing comes up in the error.log

Launching the exec file from within the .app (MyApp.app > Contents > MacOs > MyApp) works perfectly fine.

```
import logging

logging.basicConfig(filename='error.log', level=logging.ERROR)

try:
    my code

except Exception as e:
    # Log the error
    logging.error(str(e))
```
  
Launching the exec file from within the .app (MyApp.app > Contents > MacOs > MyApp) works perfectly fine.

Here is my .spec file.
```
# -*- mode: python ; coding: utf-8 -*-


a = Analysis(
    ['main.py'],
    pathex=['/Users/webmojo/PycharmProjects/WM-Final-Mac'],
    binaries=[],
    datas=[('licensing.py', '.'), ('validation.py', '.'), ('WM.icns', '.')],
    hiddenimports=[
        'ttkthemes',
        'requests',
        'py_machineid',
        'keyring',
    ],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    noarchive=False,
    py_modules=['licensing', 'validation'],
)
pyz = PYZ(a.pure, a.zipped_data)

exe = EXE(
    pyz,
    a.scripts,
    [],
    exclude_binaries=True,
    name='WM',
    debug=True,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=False,
    disable_windowed_traceback=False,
    argv_emulation=True,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
    icon=['WM.icns'],
    onefile=True,
    windowed=True,
)
coll = COLLECT(
    exe,
    a.binaries,
    a.datas,
    strip=False,
    upx=True,
    upx_exclude=[],
    name='WM',
)
app = BUNDLE(
    coll,
    name='WM.app',
    icon='WM.icns',
    bundle_identifier=None,
)
```

Any guidance and help would be much appreciated. 
Thank you!

**Comment by rokm** [comment, marked as answer]
>```
>import logging
>
>logging.basicConfig(filename='error.log', level=logging.ERROR)
>
>try:
>    my code
>
>except Exception as e:
>    # Log the error
>    logging.error(str(e))
>``` 

You seem to be trying to write your `error.log`in the current working directory. When you "open" an .app bundle, your current working directory is most likely not writable (I think it is `/` , but it could also be the temporary read-only location where the OS translocates untrusted .app bundles), so your exception handling probably raises an exception itself.

For the purposes of debugging, try to hard-code the `error.log` path to be an absolute path in your user's home directory, as was suggested in other similar issues (https://github.com/pyinstaller/pyinstaller/issues/5154#issuecomment-752203855, https://github.com/pyinstaller/pyinstaller/issues/6992#issuecomment-1198434935).

**Reply by web-mojo** [reply]
Amazing, thank you so much for your fast reply!
The error.log finally worked after your nudge in the right direction. I could work out the issue from there myself. The .app is now working! Yay!
Thanks for the great support! Would like to shout you a coffee!

**Comment by ap4499** [comment]
Thank you so much for the above - this helped me to debug my problem.

For anyone else struggling with a Unix Executable that runs fine, but a .app that does not run (MacOS).

My issue was the difference in where files are located:

.app:
        bundle_dir = Path(__file__).parent
        original_folder = Path.cwd() / bundle_dir / 'contents'

Unix executable:
        bundle_dir = Path(__file__).parent
        original_folder = Path.cwd() / bundle_dir /

The only change being "contents"

System:
Mac OS 14 (14.6.1 (23G93))
Python 12.5
PyInstaller 6.10.0

**Reply by balarohitnelluri** [reply]
Hey!

I'm facing the same issue could run Unix file but not .app on Mac. could you please enlighten more on this?

**Reply by norton120** [reply]
@ap4499 can you elaborate on your solution? When you write "the only change being 'contents'" does that mean you changed the file structure of the generated assets post-bundle, or is this some change to the pyinstaller config (I can't find `bundle_dir` or `original_folder` anywhere in pyinstaller docs), or is this a change to the application code? It sounds like you've solved it, just trying to figure out what you did :)

**Reply by ap4499** [reply]
@norton120 - annoyingly, the above doesn't mean a huge amount to me now that I am reading back on it.. I am cursing myself for the poor explanation, so my apologies. I have also changed my development approach on that project, and can no longer replicate the issue.

However, from what I recall, I believe the issue is where the application assets are located - or where it is try to store/access files that have been appended. I think this situation occurs when you have used -add-data or similar to store an Asset type folder that the application uses.

In a standard development environment, if we have logo.png contained within a folder Assets, then as we all know, the path is:
`Assets/logo.png
`

When running the Unix executable, I _think_ this works out of the box, and it is effectively doing this, where bundle_dir is the location of the .app file within the OS's file system.
```Python
        bundle_dir = Path(__file__).parent
        logo_location = Path.cwd() / bundle_dir / 'Assets' / 'logo.png'
```

However, when the .app is created, the executable is searching for Assets in what is effectively the wrong place, and it needed amending to:

```Python
# We obtain the path of the .app package and store it in the variable bundle_dir.
        bundle_dir = Path(__file__).parent

# We obtain the path to the folder where our assets will have been stored, I believe in the .app bundle, it is within a folder called contents, however I can no longer verify this.
# This is where the 'contents' part comes in
        base_path = Path.cwd() / bundle_dir / 'contents' 
```

To create a file path to access logo.png:
`logo_location = os.path.join(base_path, 'Assets', 'logo.png')`


In my applications, having experienced similar things in the past, I always have a central script that takes care of obtaining file paths. 

Say, in helper.py - I use a class, but a simplified approach would be:

```Python
def asset_path(filename):
    bundle_dir = Path(__file__).parent
    base_path = Path.cwd() / bundle_dir / 'contents' 
    return os.path.join(base_path,filename)
```

**Reply by norton120** [reply]
Thank you for the follow up, that helps a lot! I'll figure out where I want to make the path changes now that I understand the problem. Thanks again for such a thorough response!

**Reply by rokm** [reply]
In the frozen application `bundle_dir = Path(__file__).parent` should end up pointing to the "top-level application directory" (same as the PyInstaller-specific `sys._MEIPASS` variable).

In regular POSIX onedir build, this will be `/path/to/dist/<name>/_internal` directory.
In regular POSIX onefile build, this will point to the temporary directory where application extracted its files.

In (onedir-based) .app bundle, this will be `/path/to/dist/<name>.app/Contents/Frameworks` directory (the root location of `<name.app>` might change to a temporary directory if the OS decides to translocate the .app bundle to a read-only temporary location as part of its protection mechanisms).

The data you add via `--add-data` is added to this top-level application directory (actually, with .app bundle, it is added to `Contents/Resources` directory due to signing requirements, and cross-linked into `Frameworks` directory) if destination is specified as ˙'.'˙ or specified subdirectory path if you specify one in the destination part.

So having to append `contents` to the path (`base_path = Path.cwd() / bundle_dir / 'contents'`) makes little sense to me, unless you are also changing the destination part of `--add-data` (or `datas` list in the spec file) when building the .app bundle.

**Reply by rokm** [reply]
Also, this

>In a standard development environment, if we have logo.png contained within a folder Assets, then as we all know, the path is:
Assets/logo.png 

is only true if you are running your script from the directory that it is located in (which unfortunately is indeed what constitutes as "standard development environment", making people unaware of the issue with relative paths, which implicitly make assumptions about the current working directory). 

Try running it from an arbitrary directory, and it will fall apart. And the same happens when you run your frozen application when using relative paths - since current working directory might be an arbitrary one (and for certain types of builds, it will never be the top-level application directory, where your data is located).

Which is why you need to anchor your paths to `os.path.dirname(__file__)` or `pathlib.Path(__file__).parent`, in order to make your code invariant to current working directory.

**Comment by sbalk** [comment]
You can also use this workaround https://github.com/pyinstaller/pyinstaller/issues/5154#issuecomment-1567603461. Make sure you rename the path in the launcher script accordingly.
